src/bip32.c

```diff
@@ -629,14 +627,34 @@ int bip32_key_from_parent_path(const struct ext_key *hdkey,
     if (!hdkey || !child_path || !child_path_len || child_path_len > BIP32_PATH_MAX_LEN || !key_out)
         return WALLY_EINVAL;
 
+    if (flags & BIP32_FLAG_KEY_PUBLIC) {
+        /* Public derivation: Check for intermediate hardened keys */
+        for (i = 0; i < child_path_len; ++i) {
+            if (child_is_hardened(child_path[i]))
+                private_until = i + 1; /* Derive privately until this index */
+        }
+        if (private_until && !key_is_private(hdkey))
+            return WALLY_EINVAL; /* Unsupported derivation */
+    }
+
     for (i = 0; i < child_path_len; ++i) {
         struct ext_key *derived = &tmp[tmp_idx];
+        uint32_t derivation_flags = flags;
+
+        if (private_until && i < private_until - 1) {
+            /* Derive privately until we reach the last hardened child */
+            derivation_flags &= ~BIP32_FLAG_KEY_PUBLIC;
+            derivation_flags |= BIP32_FLAG_KEY_PRIVATE;
+        }
+        if (i + 2 < child_path_len)
```

Comment:
Shouldn't this be `+ 1`?

Reply:
Note the bip32 changes are being merged separately, but no, the `+ 2` is correct. This is equivalent to `i + 1 < child_path_len - 1` (zero based index vs array size). You can verify the behaviour is correct by changing this and observing the resulting test failures.